pkg/gadgets/trace/dns/tracer/gc.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build !withoutebpf
+
+package tracer
+
+import (
+	"errors"
+	"time"
+
+	"github.com/cilium/ebpf"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	log "github.com/sirupsen/logrus"
+)
+
+// Delay between each garbage collection run.
+const garbageCollectorInterval = 1 * time.Second
+
+// Max age of query before it will be garbage collected.
+const garbageCollectorMaxQueryAge = 10 * time.Second
+
+// garbageCollector runs a background goroutine to delete old query timestamps
+// from the DNS query_map. This ensures that queries that never receive a response
+// are deleted from the map.
+type garbageCollector struct {
+	started  bool
+	doneChan chan struct{}
+	queryMap *ebpf.Map
+}
+
+func newGarbageCollector(queryMap *ebpf.Map) *garbageCollector {
+	return &garbageCollector{queryMap: queryMap}
+}
+
+func (gc *garbageCollector) start() {
+	if gc.started {
+		log.Warn("Cannot start DNS queries map garbage collector because it is already running")
+		return
+	}
+
+	log.Infof("Starting garbage collection for DNS tracer")
+	gc.doneChan = make(chan struct{}, 0)
+	go gc.runLoop()
+	gc.started = true
+}
+
+func (gc *garbageCollector) stop() {
+	if !gc.started {
+		log.Warn("Cannot stop DNS queries map garbage collector because it isn't running")
+		return
+	}
+
+	log.Infof("Stopping garbage collection for DNS tracer")
+	close(gc.doneChan)
+	gc.started = false
+}
+
+func (gc *garbageCollector) runLoop() {
+	for {
+		select {
+		case <-gc.doneChan:
+			return
+
+		default:
+			log.Debugf("Executing DNS query map garbage collection")
+			gc.collect()
+			time.Sleep(garbageCollectorInterval)
+		}
```

Comment:
would that allocate a channel on every loop? I guess we could use `time.Ticker` to create the channel once, but I wasn't sure if that had any advantages over `time.Sleep` here.

Reply:
You're right. Given that it will be executed multiple times, it is better to use the ticker. I believe the implementation is cleaner with a ticker than the sleep.

Reply:
updated to use time.Ticker